Use first-class definition context for `infer`

michaelballantyne opened the pull request.

michaelballantyne committed:
Use a first-class definition context in infer

Rather than wrapping the expression to local-expand in lambda and let-syntax in order to extend the type environment, create those bindings in a first-class definition context. This change is the first step required to allow local expansion with a stop list, as a non-empty stop list always prevents expansion through primitive binding forms like the lambda and let-syntax used previously.

This commit also fixes various places when a syntax-local-introduce was
missing or in excess that caused breakage when the context binder was
changed from use-site syntax, as in the original solution with a lambda
wrapper, to an introduced temporary, as in this solution.

michaelballantyne committed:
enforce scoping of existential type variable in open

stchang commented:
Woohoo!

Would you also mind testing this repo: https://github.com/stchang/typed-rosette

And this branch: https://github.com/stchang/macrotypes/tree/peano

michaelballantyne committed:
Provide current-var-assign with syntax binding

as well as the runtime variable binding it should expand to. This is
useful for typed-rosette's syntax parameter-like rebinding used to
update the prop environment for occurrence typing.

michaelballantyne commented:
Typed rosette passes tests with the changes in this pull request: https://github.com/stchang/typed-rosette/pull/31

The peano branch merged with these changes passes tests.
